src/tests/ftest/cart_fault_injection/pool_fault_inject.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/python
+"""
+  (C) Copyright 2020-2021 Intel Corporation.
+
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+"""
+
+from apricot import TestWithServers
+from random import randint
+from command_utils_base import CommandFailure
+from general_utils import DaosTestError
+
+
+class PoolServicesFaultInjection(TestWithServers):
+    """Test class Fault domains
+    :avocado: recursive
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize a PoolServicesFaultInjection object."""
+        super().__init__(*args, **kwargs)
+        self.failed_requests = 0
+        self.object_class = self.params.get("object_class", "/run/*")
+        self.container_namespace = self.params.get("container", "/run/*")
+        self.number_servers = len(self.hostlist_servers) - 1
+
+    def look_missed_request(self, cmd_stderr, msg=b"MS request error"):
+        """ Read dmg_stderr for the msg string
+        If found, the instance attribute self.failed_request is
+        increased by 1.
+        """
+        if msg in cmd_stderr:
+            self.failed_requests += 1
+
+    def create_and_write_into_container(self):
+        """ Create a container
+        Write objects into container
+        Close container
+        Look for missed requests shown by daos command.
+        """
+        # Get and write into container
+        self.container = self.get_container(self.pool,
+                                            namespace=self.container_namespace)
+        self.look_missed_request(self.container.daos.result.stderr)
+
+        # This is done through IOR
+        # No daos/dmg message to catch
+        for server in range(0, self.number_servers):
+            self.container.write_objects(server, obj_class=self.object_class)
+        # Close container to avoid problems when deleting
+        # The container and/or pool
+        self.container.close()
+        self.look_missed_request(self.container.daos.result.stderr)
+
+    def exclude_and_reintegrate(self):
+        """ Exclude a random server from the pool
+        Wait for rebuild
+        Reintegrate the server back
+        Wait for rebuild
+
+        Due to the nature of how wait_for_rebuild() is coded
+        we can only get the last dmg command output.
+        """
+        server_to_exclude = randint(0, len(self.hostlist_servers) - 1)
+        self.pool.exclude([server_to_exclude])
+        self.look_missed_request(self.pool.dmg.result.stderr)
+
+        self.pool.wait_for_rebuild(True)
+        self.look_missed_request(self.pool.dmg.result.stderr)
+        self.pool.wait_for_rebuild(False)
+        self.look_missed_request(self.pool.dmg.result.stderr)
+        self.pool.reintegrate(str(server_to_exclude))
+        self.pool.wait_for_rebuild(True)
+        self.look_missed_request(self.pool.dmg.result.stderr)
+        self.pool.wait_for_rebuild(False)
+        self.look_missed_request(self.pool.dmg.result.stderr)
+
+    def _clean(self):
```

Comment:
You're calling this method from the main test method, so not sure why it has underscore. i.e., private

Reply:
It's private to the class so with underscore is correct.

Reply:
Yes, just want to send the message the function is only planned to be used within the class.